src/main/java/gregtech/api/metatileentity/MetaTileEntity.java

```diff
@@ -512,26 +485,38 @@ public final void onCoverLeftClick(EntityPlayer playerIn, CuboidRayTraceResult r
      */
     public boolean onRightClick(EntityPlayer playerIn, EnumHand hand, EnumFacing facing,
                                 CuboidRayTraceResult hitResult) {
+        World world = getWorld();
+        if (world == null) return false;
+
         ItemStack heldStack = playerIn.getHeldItem(hand);
         if (this instanceof IDataStickIntractable dsi) {
             if (MetaItems.TOOL_DATA_STICK.isItemEqual(heldStack) && dsi.onDataStickRightClick(playerIn, heldStack)) {
                 return true;
             }
         }
 
-        if (!playerIn.isSneaking() && openGUIOnRightClick()) {
-            if (getWorld() != null && !getWorld().isRemote) {
-                if (usesMui2()) {
-                    MetaTileEntityGuiFactory.open(playerIn, this);
-                } else {
-                    MetaTileEntityUIFactory.INSTANCE.openUI(getHolder(), (EntityPlayerMP) playerIn);
+        if (!playerIn.isSneaking()) {
+            if (!world.isRemote) {
+                EntityPlayerMP playerMP = (EntityPlayerMP) playerIn;
+                boolean uiOpened = false;
+
+                if (this instanceof IMetaTileEntityGuiHolder guiHolder &&
+                        guiHolder.shouldOpenUI(playerMP, hand, facing, hitResult)) {
+                    MetaTileEntityGuiFactory.open(playerMP, (MetaTileEntity & IMetaTileEntityGuiHolder) this);
+                    uiOpened = true;
+                } else if (openGUIOnRightClick()) {
+                    MetaTileEntityUIFactory.INSTANCE.openUI(getHolder(), playerMP);
+                    uiOpened = true;
                 }
 
-                if (getOwner() == null) {
-                    this.owner = playerIn.getUniqueID();
+                if (uiOpened) {
+                    if (getOwner() == null) {
+                        this.owner = playerIn.getUniqueID();
+                    }
+
+                    return true;
```

Comment:
This `return true` needs to move out two sets of `}` outside of the `!world.isRemote` check. Otherwise the players arm does not swing when opening the GUI, and the phantom block behavior happens